components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt

```diff
@@ -0,0 +1,413 @@
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_GEOLOCATION
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_MEDIA_KEY_SYSTEM_ACCESS
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_PERSISTENT_STORAGE
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_TRACKING
+import org.mozilla.geckoview.StorageController
+import org.mozilla.geckoview.StorageController.ClearFlags
+
+/**
+ * A storage to save [SitePermissions] using GeckoView APIs.
+ */
+@Suppress("LargeClass")
+class GeckoSitePermissionsStorage(
+    runtime: GeckoRuntime,
+    private val onDiskStorage: SitePermissionsStorage
+) : SitePermissionsStorage {
+
+    private val geckoStorage: StorageController = runtime.storageController
+    private val mainScope = CoroutineScope(Dispatchers.Main)
+
+    /*
+    * At the moment gecko view doesn't handle temporary permission,
+    * we have to store them in memory, clear them manually,
+    * until we have an API for it see:
+    * https://bugzilla.mozilla.org/show_bug.cgi?id=1710447
+    * */
+    @VisibleForTesting
+    internal val geckoTemporaryPermissions = mutableListOf<ContentPermission>()
+
+    override suspend fun save(sitePermissions: SitePermissions, request: PermissionRequest?) {
+        val geckoSavedPermissions = updateGeckoPermissionIfNeeded(sitePermissions, request)
+        onDiskStorage.save(geckoSavedPermissions, request)
+    }
+
+    override fun saveTemporary(request: PermissionRequest?) {
+        if (request is GeckoPermissionRequest.Content) {
+            geckoTemporaryPermissions.add(request.geckoPermission)
+        }
+    }
+
+    override fun clearTemporaryPermissions() {
+        geckoTemporaryPermissions.forEach {
+            geckoStorage.setPermission(it, VALUE_PROMPT)
+        }
+        geckoTemporaryPermissions.clear()
+    }
+
+    override suspend fun update(sitePermissions: SitePermissions) {
+        val updatedPermission = updateGeckoPermissionIfNeeded(sitePermissions)
+        onDiskStorage.update(updatedPermission)
+    }
+
+    override suspend fun findSitePermissionsBy(origin: String, includeTemporary: Boolean): SitePermissions? {
+        /**
+         * GeckoView ony persists [GeckoPermissionRequest.Content] other permissions like
+         * [GeckoPermissionRequest.Media], we have to store them ourselves.
+         * For this reason, we query both storage ([geckoStorage] and [onDiskStorage]) and
+         * merge both results into one [SitePermissions] object.
+         */
+        val onDiskPermission: SitePermissions? = onDiskStorage.findSitePermissionsBy(origin)
+        val geckoPermissions = findGeckoContentPermissionBy(origin, includeTemporary).groupByType()
+
+        return mergePermissions(onDiskPermission, geckoPermissions)
+    }
+
+    override suspend fun getSitePermissionsPaged(): DataSource.Factory<Int, SitePermissions> {
```

Comment:
Do we need this paging API in Fenix?

Reply:
Yes, we need it for showing the list of permissions exceptions see https://github.com/mozilla-mobile/fenix/pull/19553/files#diff-d081cee3926a40933f15182c8a61ecf824345e867e4e0499624efd3d131ebf9fR62

 **Settings -> Site permissions -> Exceptions page**
<img width="613" alt="image" src="https://user-images.githubusercontent.com/773158/118834455-3c377180-b890-11eb-8f42-3b745ab1de46.png">